File: Engine/Sources/SorobanEngine/Sheet/SheetStore.swift
import Anzan
import Foundation

/// Shared by every sheet in a store. Two jobs:
///  1. The owning-sheet stack — while a formula on sheet X evaluates, its
///     unqualified `A:1` references must resolve against X, not whichever
///     tab is active in the UI.
///  2. Cross-sheet cycle detection — `Sheet1!A:1 → Sheet2!B:1 → Sheet1!A:1`
///     must report a circular reference, not recurse forever, so the
///     in-flight set is keyed by (sheet identity, address).
final class ResolutionContext {
    struct CellKey: Hashable {
        let sheet: ObjectIdentifier
        let address: CellAddress
    }

    var resolving: Set<CellKey> = []

    private var sheetStack: [Spreadsheet] = []
    private var keyStack: [CellKey] = []
    var currentSheet: Spreadsheet? { sheetStack.last }
    /// The cell whose formula is evaluating right now — dependency edges
    /// point at it.
    var currentKey: CellKey? { keyStack.last }

    func push(_ sheet: Spreadsheet, evaluating key: CellKey) {
        sheetStack.append(sheet)
        keyStack.append(key)
    }

    func pop() {
        sheetStack.removeLast()
        keyStack.removeLast()
    }

    // MARK: Dependency graph (edit → invalidate only the affected closure)

    /// source cell → cells whose formulas read it.
    private var dependents: [CellKey: Set<CellKey>] = [:]

    /// Range reads, per source sheet: rect + reader.
    private struct RangeRead: Hashable {
        let rows: ClosedRange<Int>
        let columns: ClosedRange<Int>
        let reader: CellKey
    }
    private var rangeDependents: [ObjectIdentifier: Set<RangeRead>] = [:]

    /// Registered sheets, weakly — invalidation needs to reach their memos.
    private var sheetAccessors: [ObjectIdentifier: () -> Spreadsheet?] = [:]

    func register(_ sheet: Spreadsheet) {
        sheetAccessors[ObjectIdentifier(sheet)] = { [weak sheet] in sheet }
    }

    /// Called when a formula reads one cell of `source`.
    func recordCellRead(of source: CellKey) {
        guard let reader = currentKey, reader != source else { return }
        dependents[source, default: []].insert(reader)
    }

    /// Called when a formula reads a rectangle of `sheet`.
    func recordRangeRead(sheet: ObjectIdentifier,
                         rows: ClosedRange<Int>, columns: ClosedRange<Int>) {
        guard let reader = currentKey else { return }
        rangeDependents[sheet, default: []].insert(
            RangeRead(rows: rows, columns: columns, reader: reader))
    }

    /// A cell changed: drop its memo and, transitively, every reader's —
    /// across sheets. Edges may be stale (a reader's formula changed since
    /// recording); that only over-invalidates, which is correctness-safe.
    func invalidate(_ start: CellKey) {
        var queue = [start]
        var visited: Set<CellKey> = []
        while let key = queue.popLast() {
            guard visited.insert(key).inserted else { continue }
            sheetAccessors[key.sheet]?()?.clearMemo(at: key.address)
            if let direct = dependents[key] {
                queue.append(contentsOf: direct)
            }
            if let ranges = rangeDependents[key.sheet] {
                for read in ranges
                where read.rows.contains(key.address.row)
                    && read.columns.contains(key.address.column) {
                    queue.append(read.reader)
                }
            }
        }
    }

    /// Everything is suspect (variables changed, sheets renamed/removed,
    /// workbook loaded): clear all memos and start the graph fresh.
    func invalidateEverything() {
        dependents.removeAll()
        rangeDependents.removeAll()
        for accessor in sheetAccessors.values {
            accessor()?.clearAllMemo()
        }
    }
}

/// One worksheet: a calculation grid plus its name and layout — or, when
/// `data` is set, a DATA sheet backed by a DataStore table (the grid then
/// exists but stays empty; edits write to the table, within its bounds).
public final class Sheet {
    public internal(set) var name: String
    public let grid: Spreadsheet
    public internal(set) var data: DataSheet?
    /// Sparse non-default sizes, in points (the app clamps).
    public var columnWidths: [Int: Double] = [:]
    public var rowHeights: [Int: Double] = [:]
    /// Sparse per-cell presentation (display-only — never touches the
    /// dependency graph). Defaults are pruned, not stored; empty cells may
    /// be formatted (fill a region before its data arrives).
    public var formats: [CellAddress: CellFormat] = [:]

    public var isData: Bool { data != nil }

    init(name: String, grid: Spreadsheet, data: DataSheet? = nil) {
        self.name = name
        self.grid = grid
        self.data = data
        grid.displayName = name
    }
}

/// An ordered collection of named worksheets sharing one Calculator.
/// Owns the resolver wiring: qualified references (`Budget!A:1`) route by
/// name; unqualified ones go to the formula's owning sheet, falling back to
/// the active sheet (the log's perspective).
public final class SheetStore {
    public static let maxSheets = 256
    public static let maxNameLength = 128

    public private(set) var sheets: [Sheet] = []
    public var activeIndex = 0 {
        didSet { activeIndex = min(max(activeIndex, 0), sheets.count - 1) }
    }

    public var activeSheet: Sheet { sheets[activeIndex] }

    private let calculator: Calculator
    private let context = ResolutionContext()

    /// The calculation log, for the `History` reflection API (log-only). Set by
    /// the host (the app's `CalculatorSession` via a small adapter); nil in the
    /// CLI/tests without a log, where `History` is simply unknown. Strong: the
    /// adapter back-references the session weakly, so there's no cycle.
    public var logSource: (any LogSource)?

    public init(calculator: Calculator) {
        self.calculator = calculator
        sheets = [Sheet(name: "Sheet 1",
                        grid: Spreadsheet(calculator: calculator, context: context))]

        calculator.cellResolver = { [weak self] sheetName, column, row in
            guard let self else { throw EngineError.domainError(message: "no sheets available") }
            // Unqualified refs inside a grid formula belong to the owning
            // grid (data sheets never own formulas).
            if sheetName == nil, let current = self.context.currentSheet {
                return try current.numericValue(column: column, row: row)
            }
            let target = try self.sheet(forReference: sheetName)
            if let data = target.data {
                return try data.numericValue(column: column, row: row)
            }
            return try target.grid.numericValue(column: column, row: row)
        }
        calculator.rangeResolver = { [weak self] sheetName, fc, fr, tc, tr in
            guard let self else { throw EngineError.domainError(message: "no sheets available") }
            if sheetName == nil, let current = self.context.currentSheet {
                return try current.numericValues(fromColumn: fc, fromRow: fr,
                                                 toColumn: tc, toRow: tr)
            }
            let target = try self.sheet(forReference: sheetName)
            if let data = target.data {
                return try data.numericValues(fromColumn: fc, fromRow: fr,
                                              toColumn: tc, toRow: tr)
            }
            return try target.grid.numericValues(fromColumn: fc, fromRow: fr,
                                                 toColumn: tc, toRow: tr)
        }

        // Named cells: 'Projected Rate' routes like an unqualified A:1
        // (owning sheet, active from the log); Budget!'Rate' by sheet name.
        calculator.nameResolver = { [weak self] sheetName, name in
            guard let self else { throw EngineError.domainError(message: "no sheets available") }
            if sheetName == nil, let current = self.context.currentSheet {
                return try current.numericValue(forName: name)
            }
            let target = try self.sheet(forReference: sheetName)
            guard !target.isData else {
                throw EngineError.domainError(
                    message: "data sheets don't have named cells")
            }
            return try target.grid.numericValue(forName: name)
        }

        // Sheet-scoped λ/𝑖 definitions: resolved against the formula's
        // owning sheet (mid-evaluation) or the active tab (log input).
        calculator.scopedFunctionResolver = { [weak self] name in
            guard let self else { return nil }
            return self.scopeSheet.definedFunction(named: name)
        }
        calculator.scopedVariableResolver = { [weak self] name in
            guard let self else { return nil }
            return try self.scopeSheet.definedValue(named: name)
        }
        calculator.scopedDefinitionOwner = { [weak self] name in
            guard let self else { return nil }
            return self.scopeSheet.definitionOwner(named: name)
        }
        calculator.scopedDataTypeResolver = { [weak self] name in
            guard let self else { return nil }
            return self.scopeSheet.definedDataType(named: name)
        }

        installReflection()
        installMutation()
    }

    /// Wires the read-only Workbook reflection API: the `Workbook` global and
    /// the flat `cell()`/`sheetName()`/`sheetNames()`/`rowCount()`/`columnCount()`
    /// accessors. All hand back `.host(…)` handles (or plain values) the
    /// language navigates uniformly. The CLI leaves these unwired — `Workbook`
    /// and `cell()` are simply unknown there.
    private func installReflection() {
        calculator.hostValueResolver = { [weak self] name, inLog in
            guard let self else { return nil }
            // Reflection names are case-sensitive, like data types/constructors.
            switch name {
            case "Workbook":
                return .host(WorkbookObject(store: self))
            case "History":
                // Log-only: the array on the log path, nil in a cell (where the
                // name then degrades to a text label, not an error).
                guard inLog, let source = self.logSource else { return nil }
                return HistoryReflection.value(from: source)
            default:
                return nil
            }
        }
        calculator.hostFunctionResolver = { [weak self] name, arguments in
            guard let self else { return nil }
            switch name {
            case "cell":
                // cell(col, row) on the scope sheet; cell(sheet, col, row) by name.
                return try self.reflectCell(arguments)
            case "sheetNames":
                try Self.expectArity(arguments, 0, name)
                return .array(self.sheets.map { .string($0.name) })
            case "sheetName":
                try Self.expectArity(arguments, 0, name)
                return .string(self.scopeSheetItem.name)
            case "rowCount":
                try Self.expectArity(arguments, 0, name)
                return .number(BigDecimal(Spreadsheet.rowCount))
            case "columnCount":
                try Self.expectArity(arguments, 0, name)
                return .number(BigDecimal(Spreadsheet.columnCount))
            default:
                return nil // not a reflection function — fall through to unknown
            }
        }
    }

    /// Wires the DEFAULT (direct, no-undo) workbook mutation API:
    /// `updateCell` / `addWorksheet` / `renameWorksheet` / `deleteWorksheet`.
    /// These change the workbook, so they run from the LOG only — `inLog` is
    /// false during cell recalc and the resolver throws then (recalc stays
    /// reproducible). The app OVERRIDES `hostMutationResolver` afterward to make
    /// the same commands undoable; this default is what the CLI/tests see.
    /// Implementations live in `SheetStore+Mutation.swift`.
    private func installMutation() {
        calculator.hostMutationResolver = { [weak self] name, arguments, inLog in
            guard let self else { return nil }
            guard Self.mutationNames.contains(name) else { return nil }
            guard inLog else {
                throw EngineError.domainError(message:
                    "'\(name)' changes the workbook — it runs in the calculation log, not a cell")
            }
            switch name {
            case "updateCell": return try self.mutateUpdateCell(arguments)
            case "addWorksheet": return try self.mutateAddWorksheet(arguments)
            case "renameWorksheet": return try self.mutateRenameWorksheet(arguments)
            case "deleteWorksheet": return try self.mutateDeleteWorksheet(arguments)
            default: return nil
            }
        }
    }

    /// The mutation command names — shared by the engine default and the app
    /// override so both gate the same set log-only.
    public static let mutationNames: Set<String> =
        ["updateCell", "addWorksheet", "renameWorksheet", "deleteWorksheet"]

    /// The Sheet whose definitions/grid are in scope right now (owning sheet
    /// mid-evaluation, active tab from the log) — the Sheet-level companion to
    /// `scopeSheet`, which returns its grid.
    private var scopeSheetItem: Sheet {
        if let current = context.currentSheet {
            return sheets.first { $0.grid === current } ?? activeSheet
        }
        return activeSheet
    }

    /// `cell(col, row)` on the scope sheet, or `cell(sheet, col, row)` by name.
    private func reflectCell(_ arguments: [Value]) throws -> Value {
        switch arguments.count {
        case 2:
            return try CellObject.make(on: scopeSheet, arguments: arguments)
        case 3:
            guard case .string(let sheetName) = arguments[0] else {
                throw EngineError.domainError(
                    message: "cell()'s first argument is a sheet name — cell(\"Budget\", \"A\", 1)")
            }
            guard let sheet = sheet(named: sheetName) else {
                throw EngineError.domainError(message: "unknown sheet '\(sheetName)'")
            }
            return try CellObject.make(on: sheet.grid, arguments: Array(arguments[1...]))
        default:
            throw EngineError.domainError(
                message: "cell() takes (column, row) or (sheet, column, row)")
        }
    }

    private static func expectArity(_ arguments: [Value], _ count: Int, _ name: String) throws {
        guard arguments.count == count else {
            throw EngineError.domainError(
                message: "\(name)() takes \(count) argument\(count == 1 ? "" : "s")")
        }
    }

    /// The grid whose definitions are in scope right now.
    private var scopeSheet: Spreadsheet {
        context.currentSheet ?? activeSheet.grid
    }

    /// Where a reference points: named sheet, else the active one (log input).
    private func sheet(forReference sheetName: String?) throws -> Sheet {
        guard let sheetName else { return activeSheet }
        guard let sheet = sheet(named: sheetName) else {
            throw EngineError.domainError(message: "unknown sheet '\(sheetName)'")
        }
        return sheet
    }

    public func sheet(named name: String) -> Sheet? {
        sheets.first { $0.name.compare(name, options: .caseInsensitive) == .orderedSame }
    }

    // MARK: Structure

    @discardableResult
    public func addSheet() throws -> Sheet {
        guard sheets.count < Self.maxSheets else {
            throw EngineError.domainError(message: "a workbook holds at most \(Self.maxSheets) sheets")
        }
        var n = sheets.count + 1
        while sheet(named: "Sheet \(n)") != nil { n += 1 }
        let sheet = Sheet(name: "Sheet \(n)",
                          grid: Spreadsheet(calculator: calculator, context: context))
        sheets.append(sheet)
        return sheet
    }

    /// Adds an empty grid sheet with a specific (validated) name — the
    /// mutation API's `addWorksheet(name)`. `addSheet()` (auto-named) stays
    /// the UI's +-button path.
    @discardableResult
    public func addSheet(named name: String) throws -> Sheet {
        guard sheets.count < Self.maxSheets else {
            throw EngineError.domainError(message: "a workbook holds at most \(Self.maxSheets) sheets")
        }
        let validated = try Self.validated(name: name, existing: sheets, exceptIndex: nil)
        let sheet = makeSheet(name: validated)
        sheets.append(sheet)
        return sheet
    }

    public func removeSheet(at index: Int) throws {
        guard sheets.count > 1 else {
            throw EngineError.domainError(message: "a workbook needs at least one sheet")
        }
        guard sheets.indices.contains(index) else { return }
        sheets.remove(at: index)
        activeIndex = min(activeIndex, sheets.count - 1)
        recalculate() // formulas referencing the removed sheet become errors
    }

    public func rename(at index: Int, to newName: String) throws {
        guard sheets.indices.contains(index) else { return }
        let name = try Self.validated(name: newName, existing: sheets, exceptIndex: index)
        sheets[index].name = name
        sheets[index].grid.displayName = name
        recalculate() // references resolve by name
    }

    /// Trimmed, non-empty, ≤128 chars, unique (case-insensitive), and free of
    /// the characters that would break the `Sheet!A:1` syntax.
    public static func validated(name: String, existing: [Sheet], exceptIndex: Int?) throws -> String {
        let trimmed = name.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else {
            throw EngineError.domainError(message: "sheet names can't be empty")
        }
        guard trimmed.count <= maxNameLength else {
            throw EngineError.domainError(message: "sheet names are limited to \(maxNameLength) characters")
        }
        guard !trimmed.contains("!"), !trimmed.contains("'") else {
            throw EngineError.domainError(message: "sheet names can't contain ! or '")
        }
        for (i, sheet) in existing.enumerated() where i != exceptIndex {
            if sheet.name.compare(trimmed, options: .caseInsensitive) == .orderedSame {
                throw EngineError.domainError(message: "a sheet named '\(trimmed)' already exists")
            }
        }
        return trimmed
    }

    /// Drops every sheet's memo — a log variable changed, a sheet was
    /// renamed/removed, or a workbook loaded.
    public func recalculate() {
        for sheet in sheets {
            sheet.grid.recalculate()
        }
    }

    /// Replaces everything (workbook open / new).
    public func replaceSheets(_ newSheets: [Sheet], activeName: String?) {
        precondition(!newSheets.isEmpty)
        sheets = newSheets
        activeIndex = newSheets.firstIndex {
            $0.name.compare(activeName ?? "", options: .caseInsensitive) == .orderedSame
        } ?? 0
        recalculate()
    }

    /// A fresh empty sheet built against this store's shared context —
    /// for workbook loading.
    public func makeSheet(name: String) -> Sheet {
        Sheet(name: name, grid: Spreadsheet(calculator: calculator, context: context))
    }

    /// A DATA sheet over a DataStore table (for loading/import).
    public func makeDataSheet(name: String, data: DataSheet) -> Sheet {
        Sheet(name: name, grid: Spreadsheet(calculator: calculator, context: context), data: data)
    }

    /// Imports a data sheet at the end (name pre-validated by the caller's
    /// uniquifier or validated here).
    @discardableResult
    public func addDataSheet(named name: String, data: DataSheet) throws -> Sheet {
        guard sheets.count < Self.maxSheets else {
            throw EngineError.domainError(message: "a workbook holds at most \(Self.maxSheets) sheets")
        }
        let validated = try Self.validated(name: name, existing: sheets, exceptIndex: nil)
        let sheet = makeDataSheet(name: validated, data: data)
        sheets.append(sheet)
        recalculate()
        return sheet
    }
}

File: rust/engine/src/sheet_store.rs
//! An ordered collection of named worksheets sharing one Calculator.
//! Owns the resolver wiring: qualified references (`Budget!A:1`) route by
//! name; unqualified ones go to the formula's owning sheet, falling back to
//! the active sheet (the log's perspective).
//!
//! Ownership: the store holds the Calculator in an `Rc<RefCell<…>>`; the
//! resolver closures installed INTO the calculator capture a `Weak` of the
//! store's internals and receive the evaluator + environment as call
//! arguments, so they never touch the Calculator RefCell — the only borrow
//! of it is the host's outermost call.

use crate::cell_address::CellAddress;
use crate::context::ResolutionContext;
use crate::data_store::DataSheet;
use crate::spreadsheet::{CellDisplay, Spreadsheet};
use anzan::{BigDecimal, Calculator, EngineError, Value};
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

/// One worksheet: a calculation grid plus its name and layout — or, when
/// `data` is set, a DataStore-backed table (CSV-imported, SQLite-persisted).
/// A data sheet still carries a `grid` (for the display name / shared context)
/// but its cell reads route to the table instead; it owns no formulas.
pub struct Sheet {
    pub(crate) name: RefCell<String>,
    pub grid: Rc<Spreadsheet>,
    /// Sparse non-default sizes, in points (the app clamps).
    pub column_widths: RefCell<HashMap<usize, f64>>,
    pub row_heights: RefCell<HashMap<usize, f64>>,
    /// Sparse per-cell presentation (display-only — never touches the
    /// dependency graph). Defaults are pruned, not stored; empty cells may
    /// be formatted (fill a region before its data arrives).
    pub formats: RefCell<HashMap<CellAddress, crate::cell_format::CellFormat>>,
    /// When set, this is a data sheet: reads resolve against the table (bounded
    /// by the table, not the 1000-row grid), and it owns no formulas.
    pub data: RefCell<Option<DataSheet>>,
}

impl Sheet {
    fn new(name: &str, grid: Rc<Spreadsheet>) -> Rc<Self> {
        grid.set_display_name(Some(name.to_string()));
        Rc::new(Self {
            name: RefCell::new(name.to_string()),
            grid,
            column_widths: RefCell::new(HashMap::new()),
            row_heights: RefCell::new(HashMap::new()),
            formats: RefCell::new(HashMap::new()),
            data: RefCell::new(None),
        })
    }

    pub fn name(&self) -> String {
        self.name.borrow().clone()
    }

    /// True when this sheet is backed by a DataStore table rather than a grid.
    pub fn is_data(&self) -> bool {
        self.data.borrow().is_some()
    }
}

pub(crate) struct StoreInner {
    pub(crate) sheets: RefCell<Vec<Rc<Sheet>>>,
    pub(crate) active_index: std::cell::Cell<usize>,
    pub(crate) context: Rc<ResolutionContext>,
    /// The calculation log, for the `History` reflection API (log-only).
    /// Set by the host; `None` in the CLI/tests without a log, where
    /// `History` is simply unknown.
    pub(crate) log_source: RefCell<Option<Rc<dyn crate::history_reflection::LogSource>>>,
}

impl StoreInner {
    pub(crate) fn active_sheet(&self) -> Rc<Sheet> {
        let sheets = self.sheets.borrow();
        let index = self.active_index.get().min(sheets.len() - 1);
        Rc::clone(&sheets[index])
    }

    pub(crate) fn sheet_named(&self, name: &str) -> Option<Rc<Sheet>> {
        let needle = name.to_lowercase();
        self.sheets
            .borrow()
            .iter()
            .find(|s| s.name().to_lowercase() == needle)
            .cloned()
    }

    /// Where a reference points: named sheet, else the active one (log
    /// input).
    pub(crate) fn sheet_for_reference(
        &self,
        sheet_name: Option<&str>,
    ) -> Result<Rc<Sheet>, EngineError> {
        let Some(sheet_name) = sheet_name else {
            return Ok(self.active_sheet());
        };
        self.sheet_named(sheet_name)
            .ok_or_else(|| EngineError::domain(format!("unknown sheet '{sheet_name}'")))
    }

    /// The grid whose definitions are in scope right now (owning sheet
    /// mid-evaluation, active tab from the log).
    pub(crate) fn scope_sheet(&self) -> Rc<Spreadsheet> {
        self.context
            .current_sheet()
            .unwrap_or_else(|| Rc::clone(&self.active_sheet().grid))
    }

    /// The Sheet whose definitions/grid are in scope right now — the
    /// Sheet-level companion to `scope_sheet`, which returns its grid.
    pub(crate) fn scope_sheet_item(&self) -> Rc<Sheet> {
        if let Some(current) = self.context.current_sheet() {
            let sheets = self.sheets.borrow();
            if let Some(sheet) = sheets.iter().find(|s| Rc::ptr_eq(&s.grid, &current)) {
                return Rc::clone(sheet);
            }
        }
        self.active_sheet()
    }

    // MARK: Sheet mechanics
    //
    // These live on StoreInner (not SheetStore) so the resolver closures —
    // which capture only a `Weak<StoreInner>` — can mutate the store. The
    // SheetStore methods of the same names delegate here.

    /// Drops every sheet's memo — a log variable changed, a sheet was
    /// renamed/removed, or a workbook loaded.
    pub(crate) fn recalculate(&self) {
        for sheet in self.sheets.borrow().iter() {
            sheet.grid.recalculate();
        }
    }

    /// A fresh empty sheet built against this store's shared context.
    pub(crate) fn make_sheet(&self, name: &str) -> Rc<Sheet> {
        Sheet::new(name, Spreadsheet::new(Rc::clone(&self.context)))
    }

    /// A data sheet: an ordinary sheet shell whose reads route to `data`.
    pub(crate) fn make_data_sheet(&self, name: &str, data: DataSheet) -> Rc<Sheet> {
        let sheet = self.make_sheet(name);
        *sheet.data.borrow_mut() = Some(data);
        sheet
    }

    fn check_capacity(&self) -> Result<(), EngineError> {
        if self.sheets.borrow().len() >= SheetStore::MAX_SHEETS {
            return Err(EngineError::domain(format!(
                "a workbook holds at most {} sheets",
                SheetStore::MAX_SHEETS
            )));
        }
        Ok(())
    }

    /// Adds an empty grid sheet with a specific (validated) name — the
    /// mutation API's `addWorksheet(name)`.
    pub(crate) fn add_sheet_named(&self, name: &str) -> Result<Rc<Sheet>, EngineError> {
        self.check_capacity()?;
        let validated = self.validated_name(name, None)?;
        let sheet = self.make_sheet(&validated);
        self.sheets.borrow_mut().push(Rc::clone(&sheet));
        Ok(sheet)
    }

    pub(crate) fn remove_sheet(&self, index: usize) -> Result<(), EngineError> {
        {
            let mut sheets = self.sheets.borrow_mut();
            if sheets.len() <= 1 {
                return Err(EngineError::domain("a workbook needs at least one sheet"));
            }
            if index >= sheets.len() {
                return Ok(());
            }
            sheets.remove(index);
            let count = sheets.len();
            self.active_index
                .set(self.active_index.get().min(count - 1));
        }
        self.recalculate(); // formulas referencing the removed sheet error
        Ok(())
    }

    pub(crate) fn rename(&self, index: usize, new_name: &str) -> Result<(), EngineError> {
        {
            let sheets = self.sheets.borrow();
            if index >= sheets.len() {
                return Ok(());
            }
            let name = self.validated_name_against(new_name, &sheets, Some(index))?;
            *sheets[index].name.borrow_mut() = name.clone();
            sheets[index].grid.set_display_name(Some(name));
        }
        self.recalculate(); // references resolve by name
        Ok(())
    }

    fn validated_name(&self, name: &str, except: Option<usize>) -> Result<String, EngineError> {
        let sheets = self.sheets.borrow();
        self.validated_name_against(name, &sheets, except)
    }

    /// Trimmed, non-empty, ≤128 chars, unique (case-insensitive), and free
    /// of the characters that would break the `Sheet!A:1` syntax.
    fn validated_name_against(
        &self,
        name: &str,
        existing: &[Rc<Sheet>],
        except: Option<usize>,
    ) -> Result<String, EngineError> {
        let trimmed = name.trim();
        if trimmed.is_empty() {
            return Err(EngineError::domain("sheet names can't be empty"));
        }
        if trimmed.chars().count() > SheetStore::MAX_NAME_LENGTH {
            return Err(EngineError::domain(format!(
                "sheet names are limited to {} characters",
                SheetStore::MAX_NAME_LENGTH
            )));
        }
        if trimmed.contains('!') || trimmed.contains('\'') {
            return Err(EngineError::domain("sheet names can't contain ! or '"));
        }
        for (i, sheet) in existing.iter().enumerate() {
            if Some(i) == except {
                continue;
            }
            if sheet.name().to_lowercase() == trimmed.to_lowercase() {
                return Err(EngineError::domain(format!(
                    "a sheet named '{trimmed}' already exists"
                )));
            }
        }
        Ok(trimmed.to_string())
    }
}

/// Workbook mutation — the log-only commands' direct (no-undo) default.
mod mutation;
/// Installs the Calculator resolver closures.
mod resolvers;

pub struct SheetStore {
    pub(crate) inner: Rc<StoreInner>,
    calculator: Rc<RefCell<Calculator>>,
}

impl SheetStore {
    pub const MAX_SHEETS: usize = 256;
    pub const MAX_NAME_LENGTH: usize = 128;

    /// The mutation command names — the log-only gated set, shared so any
    /// host override gates exactly the same commands.
    pub const MUTATION_NAMES: [&'static str; 4] = [
        "updateCell",
        "addWorksheet",
        "renameWorksheet",
        "deleteWorksheet",
    ];

    pub fn new(calculator: Rc<RefCell<Calculator>>) -> Self {
        let context = ResolutionContext::new();
        let inner = Rc::new(StoreInner {
            sheets: RefCell::new(Vec::new()),
            active_index: std::cell::Cell::new(0),
            context: Rc::clone(&context),
            log_source: RefCell::new(None),
        });
        inner
            .sheets
            .borrow_mut()
            .push(Sheet::new("Sheet 1", Spreadsheet::new(Rc::clone(&context))));

        let store = Self {
            inner: Rc::clone(&inner),
            calculator: Rc::clone(&calculator),
        };
        store.install_resolvers();
        store
    }

    pub fn calculator(&self) -> &Rc<RefCell<Calculator>> {
        &self.calculator
    }

    /// Attaches the host's log for `History` reflection (log-only).
    pub fn set_log_source(&self, source: Rc<dyn crate::history_reflection::LogSource>) {
        *self.inner.log_source.borrow_mut() = Some(source);
    }

    // MARK: Host conveniences (the outermost-borrow entry points)

    /// What the grid shows at `address` on the active sheet — the harness's
    /// and UI's read path (borrows the calculator once, here).
    pub fn display_value(&self, address: CellAddress) -> CellDisplay {
        let sheet = self.inner.active_sheet();
        self.display_value_on(&sheet, address)
    }

    pub fn display_value_on(&self, sheet: &Rc<Sheet>, address: CellAddress) -> CellDisplay {
        // A data sheet renders straight from the table — no calculator, no
        // formulas: empty → blank, parseable → a number, else the raw text
        // (headers and labels). Mirrors Swift's `SheetModel.display(at:)`.
        if let Some(data) = &*sheet.data.borrow() {
            let raw = data.raw_value(address.row, address.column);
            if raw.is_empty() {
                return CellDisplay::Empty;
            }
            return match BigDecimal::parse(&raw) {
                Some(value) => CellDisplay::Value(value),
                None => CellDisplay::Text(raw),
            };
        }
        self.calculator
            .borrow_mut()
            .host_eval(|evaluator, environment| {
                sheet.grid.display_value((evaluator, environment), address)
            })
    }

    // MARK: Sheets

    pub fn sheets(&self) -> Vec<Rc<Sheet>> {
        self.inner.sheets.borrow().clone()
    }

    pub fn active_sheet(&self) -> Rc<Sheet> {
        self.inner.active_sheet()
    }

    pub fn set_active_index(&self, index: usize) {
        let count = self.inner.sheets.borrow().len();
        self.inner
            .active_index
            .set(index.min(count.saturating_sub(1)));
    }

    /// The index of the currently active sheet (clamped into range).
    pub fn active_index(&self) -> usize {
        let count = self.inner.sheets.borrow().len();
        self.inner.active_index.get().min(count.saturating_sub(1))
    }

    /// The number of open sheets.
    pub fn sheet_count(&self) -> usize {
        self.inner.sheets.borrow().len()
    }

    pub fn sheet_named(&self, name: &str) -> Option<Rc<Sheet>> {
        self.inner.sheet_named(name)
    }

    /// Adds an auto-named empty grid sheet (the UI's +-button path).
    pub fn add_sheet(&self) -> Result<Rc<Sheet>, EngineError> {
        let mut n = self.inner.sheets.borrow().len() + 1;
        while self.sheet_named(&format!("Sheet {n}")).is_some() {
            n += 1;
        }
        self.inner.add_sheet_named(&format!("Sheet {n}"))
    }

    /// Adds an empty grid sheet with a specific (validated) name — the
    /// mutation API's `addWorksheet(name)`.
    pub fn add_sheet_named(&self, name: &str) -> Result<Rc<Sheet>, EngineError> {
        self.inner.add_sheet_named(name)
    }

    pub fn remove_sheet(&self, index: usize) -> Result<(), EngineError> {
        self.inner.remove_sheet(index)
    }

    pub fn rename(&self, index: usize, new_name: &str) -> Result<(), EngineError> {
        self.inner.rename(index, new_name)
    }

    /// Renames the sheet at `index` and rewrites every cross-sheet reference
    /// (`Old!A:1` / `'Old'!A:1`) to match — the UI tab-rename path.
    pub fn rename_worksheet(&self, index: usize, new_name: &str) -> Result<(), EngineError> {
        self.inner.rename_worksheet(index, new_name)
    }

    /// Drops every sheet's memo — a log variable changed, a sheet was
    /// renamed/removed, or a workbook loaded.
    pub fn recalculate(&self) {
        self.inner.recalculate();
    }

    // MARK: Mutation seams (public — a host's undoable override reuses them)

    /// A `Worksheet` handle for the sheet at `index` — the value the
    /// mutation commands return, built identically by the engine default
    /// and any host override.
    pub fn worksheet_handle(&self, index: usize) -> Value {
        self.inner.worksheet_handle(index)
    }

    /// Resolves a CELL handle (`cell("A", 1)` / `…cell("A", 1)`) to the
    /// index of the sheet it lives on and its address — so a host can write
    /// it undoably.
    pub fn cell_target(&self, value: &Value) -> Result<(usize, CellAddress), EngineError> {
        self.inner.cell_target(value)
    }

    /// Resolves a worksheet TARGET — a `Worksheet` handle or a name string —
    /// to its current index in the workbook.
    pub fn sheet_index_for_target(&self, value: &Value) -> Result<usize, EngineError> {
        self.inner.sheet_index_for_target(value)
    }

    /// A value as a cell's raw text: numbers become digits, strings are
    /// verbatim. Structures/functions/handles can't live in a cell.
    pub fn raw_text_from(value: &Value) -> Result<String, EngineError> {
        match value {
            Value::Number(number) => Ok(number.to_string()),
            Value::String(text) => Ok(text.clone()),
            other => Err(EngineError::domain(format!(
                "a cell holds a number or text, not {}",
                other.kind_name()
            ))),
        }
    }

    /// Replaces everything (workbook open / new).
    pub fn replace_sheets(&self, new_sheets: Vec<Rc<Sheet>>, active_name: Option<&str>) {
        assert!(!new_sheets.is_empty());
        let active = new_sheets
            .iter()
            .position(|s| active_name.is_some_and(|n| s.name().to_lowercase() == n.to_lowercase()))
            .unwrap_or(0);
        *self.inner.sheets.borrow_mut() = new_sheets;
        self.inner.active_index.set(active);
        self.recalculate();
    }

    /// A fresh empty sheet built against this store's shared context — for
    /// workbook loading.
    pub fn make_sheet(&self, name: &str) -> Rc<Sheet> {
        self.inner.make_sheet(name)
    }

    /// A DataStore-backed sheet whose reads route to `data` — for CSV import
    /// and loading a workbook that carries data sheets.
    pub fn make_data_sheet(&self, name: &str, data: DataSheet) -> Rc<Sheet> {
        self.inner.make_data_sheet(name, data)
    }
}
